include/nbl/builtin/hlsl/cpp_compat/intrinsics.hlsl

```diff
@@ -20,222 +23,158 @@ namespace nbl
-#else
-	return glm::max(a, b);
-#endif
+	return cpp_compat_intrinsics_impl::max_helper<T>::__call(a, b);
 }
 
-template<typename FloatingPoint NBL_FUNC_REQUIRES(hlsl::is_floating_point_v<FloatingPoint>)
-inline bool isnan(NBL_CONST_REF_ARG(FloatingPoint) val)
-{
-#ifdef __HLSL_VERSION
-	return spirv::isNan(val);
-#else
-	return std::isnan(val);
-#endif
-}
-
-template <typename Integer NBL_FUNC_REQUIRES(hlsl::is_integral_v<Integer>)
-inline bool isnan(Integer val)
+template<typename FloatingPoint>
+inline FloatingPoint rsqrt(FloatingPoint x)
 {
-	using AsUint = typename unsigned_integer_of_size<sizeof(Integer)>::type;
-	using AsFloat = typename float_of_size<sizeof(Integer)>::type;
-
-	AsUint asUint = bit_cast<AsUint, Integer>(val);
-	return bool((ieee754::extractBiasedExponent<Integer>(val) == ieee754::traits<AsFloat>::specialValueExp) && (asUint & ieee754::traits<AsFloat>::mantissaMask));
+	return cpp_compat_intrinsics_impl::rsqrt_helper<FloatingPoint>::__call(x);
 }
 
-template<typename FloatingPoint NBL_FUNC_REQUIRES(hlsl::is_floating_point_v<FloatingPoint>)
-inline FloatingPoint isinf(NBL_CONST_REF_ARG(FloatingPoint) val)
+template<typename Integer>
+inline Integer bitReverse(Integer val)
 {
-#ifdef __HLSL_VERSION
-	return spirv::isInf(val);
-#else
-	return std::isinf(val);
-#endif
+	return cpp_compat_intrinsics_impl::bitReverse_helper<Integer>::__call(val);
 }
 
-template<typename Integer NBL_FUNC_REQUIRES(hlsl::is_integral_v<Integer>)
-inline bool isinf(Integer val)
+template<typename T, uint16_t Bits>
+/**
+* @brief Takes the binary representation of `value` as a string of `Bits` bits and returns a value of the same type resulting from reversing the string
+*
+* @tparam T Type of the value to operate on.
+* @tparam Bits The length of the string of bits used to represent `value`.
+*
+* @param [in] value The value to bitreverse.
+*/
+inline T bitReverseAs(T val)
 {
-	using AsUint = typename unsigned_integer_of_size<sizeof(Integer)>::type;
-	using AsFloat = typename float_of_size<sizeof(Integer)>::type;
-
-	AsUint tmp = bit_cast<AsUint>(val);
-	return (tmp & (~ieee754::traits<AsFloat>::signMask)) == ieee754::traits<AsFloat>::inf;
+	return cpp_compat_intrinsics_impl::bitReverseAs_helper<T, Bits>::__call(val);
 }
 
-template<typename  T>
-inline T exp2(NBL_CONST_REF_ARG(T) val)
+template<typename T NBL_FUNC_REQUIRES(is_unsigned_v<T>)
+/**
+* @brief Takes the binary representation of `value` and returns a value of the same type resulting from reversing the string of bits as if it was `bits` long.
+* Keep in mind `bits` cannot exceed `8 * sizeof(T)`.
+*
+* @tparam T type of the value to operate on.
+*
+* @param [in] value The value to bitreverse.
+* @param [in] bits The length of the string of bits used to represent `value`.
+*/
+T bitReverseAs(T val, uint16_t bits)
 {
-#ifdef __HLSL_VERSION
-	return spirv::exp2(val);
-#else
-	return std::exp2(val);
-#endif
+	return cpp_compat_intrinsics_impl::bitReverseAs_helper<T, 0>::__call(val, bits);
```

Comment:
I think its okay to go to the same runtime/dynamic helper, because the compiler will now know the `bits` are a compile time constant if you're coming from the templated Bits non-helper

Reply:
did it this way: https://github.com/Devsh-Graphics-Programming/Nabla/blob/new_tgmath/include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl#L337-L350

Reply:
the code you linked has nothing to do with bitReverseAs probably because you didin't use a pemalink and just a branch link

Reply:
I'm just saying we don't need two `bitReverseAs` the 
```cpp
template<typename T NBL_FUNC_REQUIRES(is_unsigned_v<T>)
T bitReverseAs(T val, uint16_t bits);
```
is the only one we want